### ISSUE TITLE
Inserting tabs in block selection still does not work properly

### ISSUE THREAD

**Issue by MartinL0815** [VsVim/VsVim#2420]
If I have the following code:
```
   123456789
1: aVariableCalledA = b +
2:     c +
3:     d;
```

If I select text the line 2 and 3 in column 5 using a visual block selection,
press ``I`` to insert a tabulator, the result is:
```
   12345678901234
1: aVariableCalledA = b +
2:         c +
3:             d;
```
Expected was:
```
   1234567890
1: aVariableCalledA = b +
2:         c +
3:         d;
```
Our VS tab settings are:
- tab size: 8
- indent size: 4
- insert spaces: true

For the release 2.6.0.0 issues have been closed adressing this issue, so I'm surprised to see this still going wrong.

**Comment by ricksladkey** [comment]
I cannot come up the right combination of settings to reproduce this with VsVim 2.6.0.

I did find a different bug with the same settings:

Our VS tab settings are:

* tab size: 8
* indent size: 4
* insert spaces: true

```
   1234567890
1: aVariableCalledA = b +
2: ------->c +
3: ------->d;
```
(where `------->` represents a literal tab expanded visually to 8 spaces).

Then the same insertion produces:

```
   1234567890
1: aVariableCalledA = b +
2: ------->        c +
3: ------->       d;
```
where eight spaces were inserted on line 2 and seven spaces were inserted on line 3.

**Comment by ricksladkey** [comment]
See related issues #992 and #1345 and PR #2211.

**Comment by ricksladkey** [comment]
@MartinL0815 Can you humor me and just double-check that you are not using the old version by typing the `:version` command?  It seems unlikely but it's the only explanation that I can think of.

**Comment by MartinL0815** [comment]
Sorry Rick, but executing version shows 2.6.0.0.
If you don't belief, I'll post a screenshot.

**Comment by ricksladkey** [comment]
@MartinL0815 Thanks, Martin. No, I trust you completely, I am just stymied because I cannot reproduce the problem. I reviewed the code line by line and in the debugger and don't see how it can happen. At this point, I would be grateful if any kind soul would try to repro it as another data point.

**Comment by ricksladkey** [comment]
Whoa, I stand corrected! I repeated it once, but I cannot repeat it again. I guess `:set softtabstop=8` is not the culprit after all.

@MartinL0815 At this point, if you could repro the issue and then type `:set` and paste the output in a comment, that would help me a lot.

**Comment by ricksladkey** [comment]
For the record, copy and pasting the results of `:set` is a little wonky.  If you double click on any of the output and then type `<C-a><C-c>`, it should copy it suitable for pasting.

**Comment by MartinL0815** [comment]
@ricksladkey, here my settings, returned by set:

```
backspace="indent,eol,start"
hlsearch
incsearch
scrolloff=5
tildeop
vimrc="C:\Users\lemburg\_vsvimrc"
vimrcpaths="C:\Users\lemburg;C:\Users\lemburg\vimfiles;C:\Users\lemburg"
nowrapscan
vsvim_useeditorindent
vsvim_useeditordefaults
vsvim_useeditortab
vsvim_useeditorcommandmargin
cursorline
autoindent
expandtab
number
shiftwidth=4
list
```

Those settings are "old". In the beginning of VsVim I shared some supported settings with my Vim installation.
Do you see something important?

BTW: Thanks!

**Comment by ricksladkey** [comment]
Thanks, @MartinL0815.

After testing the issue many times, I'm now pretty sure that the repro is in fact shifted over four spaces to the right, that is:

```
   1234567890123456
1: aVariableCalledA = b +
2:         c +
3:         d;
```

```
   1234567890123456
1: aVariableCalledA = b +
2:             c +
3:                 d;
```

Can you confirm?

If so, then I know what the problem is: The first (typed) tab is treated like an indent operation that aligns to a multiple of 4 because of the `shiftwidth=4` and `vsvim_useeditortab`. The second (generated) tab is treated like a literal tab that aligns to a multiple of 8 because of `tabstop=8`.

**Comment by MartinL0815** [comment]
Yes, I can confirm this behavior.

But - is it a bug or a  wrong configuration?
Should I remove "shiftwidth" from my rc file?

**Comment by ricksladkey** [comment]
@MartinL0815 Your `shiftwidth` setting is entirely up to you. The problem is that because "Use Visual Studio Tab and Backspace" is set, the first tab is not processed by VsVim but by Visual Studio. In turn, what Visual Studio does when you type tab is dependent on the language service that is processing the tab. Most language services for Visual Studio with treat the tab key like it is an indent key. Vim on the other hand treats tab like you want to insert a tab character and it will move you to the next tab stop.

So if you want VsVim to act like vim, you should turn off "Use Visual Studio Tab and Backspace".

Starting with this buffer:

```
   1234567890123456
1: aVariableCalledA = b +
2:         c +
3:         d;
```

and issuing this command to ensure the settings are exactly how we want:

```
:set sw=4 ts=8 et novsvim_useeditortab
```

And then typing the repro steps: `<C-v>jI<Tab><Esc>` results in the following buffer:

```
   1234567890123456
1: aVariableCalledA = b +
2:                 c +
3:                 d;
```

exactly like it does in vim.

Short answer: the work-around is to turn off `vsvim_useeditortab`. Longer answer: The bug is that when `vsvim_useeditortab` is set, block inserting a tab may not do the same thing on the second and subsequent lines of the block that the language service did on the first line of the block.

I wouldn't mind having @jaredpar's opinion about this. Ideally block insert would work correctly regardless of what your `vsvim_useeditortab` setting is. However, making it work when it *is* set is a bit tricky. These seem like possible solutions:

1) We really have no idea what the language service is going to do when you type tab, so we could just insert whatever the language service inserted on the first line into the subsequent lines
2) We could use Visual Studio automation to programmatically "type" the tab on the subsequent lines so that the language service decides what tab does on *all* the lines, not just the first one
3) We could assume the language service will treat tab as "indent to the next shift width" and handle the subsequent lines accordingly
4) We could purposely ignore the `vsvim_useeditortab` during block inserts

**Comment by ricksladkey** [comment]
I've weighed the choices:

> 1. We really have no idea what the language service is going to do when you type tab, so we could just insert whatever the language service inserted on the first line into the subsequent lines

This has the disadvantage that it doesn't "fix" tabs on the second and subsequent lines the same way that it does on the first line.

> 2. We could use Visual Studio automation to programmatically "type" the tab on the subsequent lines so that the language service decides what tab does on all the lines, not just the first one

This sounds like the most "compatible" way to handle it, but it is more work than I'm prepared to invest.

> 3. We could assume the language service will treat tab as "indent to the next shift width" and handle the subsequent lines accordingly

This is easy to do and I think it handles the 90% case pretty well.

> 4. We could purposely ignore the `vsvim_useeditortab` during block inserts

This favors consistency over predictability.

**Comment by ricksladkey** [comment]
Ha, ha, the issue that refuses to be solved!

The current design doesn't allow the VsVim core to directly conditionalize its behavior on settings like `UseEditorTabAndBackspace`. For example, in the VimApp test program, those settings don't even exist.

**Comment by ricksladkey** [comment]
> 2. We could use Visual Studio automation to programmatically "type" the tab on the subsequent lines so that the language service decides what tab does on all the lines, not just the first one

This turns out to be not as hard as I had anticipated, and is in fact the best solution. We simply repeat the insertion and the insertion command infrastructure handles any embedded tabs just like it did during the first insertion. In other words, the text that gets inserted into the second and subsequent lines isn't deduced and then inserted, it is simply repeated.

I spent way too much time fretting about this issue, but it is satisfying to finally come up with a solution that works correctly no matter what the setting of `vsvim_useeditortab`.
